### ISSUE TITLE
SNESC's save state system doesn't work with gpsp

### ISSUE THREAD

**Issue by GAmethyst** [KMFDManic/NESC-SNESC-Modifications#163]
### Step 1: Describe your setup:

  * hakchi2 CE version: 3.5.0
  * retroarch used (filename):  Hakchi_Retroarch_Neo_v1_7_4a
  * hmod(s) used by the game you are having problems with (filenames): CORE_gpsp
 
### Step 2: Describe the problem:
  
  * Using SNESC's base save state system crashes the ROM similarly to when you take the cartridge out of a physical GBA while the game is running.

**Comment by nl255** [comment]
I can confirm that this will also happen with Retroarch 1.7.5 xtreme and km_gpSP.

**Comment by KMFDManic** [comment]
In your testing, make sure you use RetroArch Settings, Exit to escape from games.  Using Reset on the system itself, is not a good way to go as far as save states are concerned.  There "was" a bug with double extensions, for gpSP that BsLeNuL helped with.  Think .gba inside a .7z or .zip, which would be the double extensions.  Try running a game uncompressed.  Also, try a standard game, such as a Super Mario one, or even F-Zero.  Let me know how things turn out.  I will do a retest when I can!

**Comment by nl255** [comment]
The save state loads but as soon as it does the game is frozen and there is this awful noise.  I tested it with Mother 3 and I use the home button (which on the 8bitdo receiver I am using now is mapped to down+select) to exit if that makes any difference.

**Comment by KMFDManic** [comment]
So, you exit game, create suspend point, then resuming it gives bad sound?  Some games I have tested worked fine.  I always use RetroArch Exit, not down+select.  Mother 3 may, in fact, end up being a game that has issues.  It used to even with mGBA.  Can you try testing F-Zero or one of the Mario games, as a frame of reference?  Mother 3 I run now using mGBA, where in game and suspends both work great. gpSP I use primarily for games such as Final Fantasy IV, V, where they run less than stellar on mGBA.

**Comment by KMFDManic** [comment]
Just tested with Metroid Zero Mission "uncompressed", and suspend state loaded fine.  In Hakchi2 CE, decompress your game, then resynch/export, and you should be ok.  This appears to not work for all games, when compressed.  So, best recommended course of action is to run uncompressed, in this case.

**Comment by KMFDManic** [comment]
After more testing, myself and another can confirm that "larger" games have issues, even on PC RetroArch.  So, stick to standard sized games for gpSP, suspend point wise, recommended uncompressed.  mGBA, otherwise:)  This can be a tricky situation, as gpSP uses real BIOS, and also introduces a new set of potential conflicts.  The accuracy of FF games, though, as an example..make gpSP well worth using, too!

Closing this issue for now, as it is a "Core" issue that may indeterminately be able to be fixed.  I truly appreciate the feedback from you both, @GAmethyst @nl255 For now, follow my advice, and I will look further into things along with BsLeNuL:)

**Comment by KMFDManic** [comment]
@GAmethyst @nl255 Test saves from RetroArch, itself.  You can set these up in Input/Hot Key Binds.  These appear to function more reliably for gpSP, but USB-HOST must be used.

**Comment by nl255** [comment]
@KMFDManic : As I said, I have already tried that and save states made in RetroArch via the start+select menu work just fine.  Only the suspend saves are broken in Mother 3 (and presumably other larger games).  It might be a good idea to simply disable suspend saving in gpSP altogether until it is fixed, assuming disabling suspend saves for a single core like that is possible.

**Comment by KMFDManic** [comment]
Suspend States work for most games of standard size that I have tested with gpSP, while uncompressed.  Mother 3 is quite a bit bigger than normal.  It also has issues saving on PC RetroArch.  So, it would seem size might be a factor in this case.  I tested about 25 games with Suspends, and all resumed after power off and power on.  Larger N64 and GBA games had save issues, that we fixed for Glupen/Mupen/Parallel/mGBA.  gpSP, if I can find time, I can try to pinpoint max size it will allow for games to be to successfully save.  Definitely still worth using for the many games it does save and play well with.
